chainstate/src/detail/chainstateref/epoch_seal.rs

```diff
@@ -110,3 +111,44 @@ fn rollback_epoch_seal<S: BlockchainStorageWrite>(
 
     Ok(())
 }
+
+/// Indicates whether a block was connected or disconnected.
+/// Stores current tip height and index if necessary.
+pub enum BlockStateEventWithIndex<'a> {
+    Connect(BlockHeight, &'a BlockIndex),
+    Disconnect(BlockHeight),
+}
+
+/// Every epoch has data associated with it.
+/// On every block change check whether this data should be updated.
+pub fn update_epoch_data<S: BlockchainStorageWrite>(
+    db_tx: &mut S,
+    chain_config: &ChainConfig,
+    block_op: BlockStateEventWithIndex<'_>,
+) -> Result<(), BlockError> {
+    match block_op {
+        BlockStateEventWithIndex::Connect(tip_height, block_index) => {
+            if chain_config.is_last_block_in_epoch(&tip_height) {
+                // Consider the randomness of the last block to be the randomness of the epoch
+                if let Some(epoch_randomness) = block_index.preconnect_data().pos_randomness() {
+                    db_tx
+                        .set_epoch_data(
+                            chain_config.epoch_index_from_height(&tip_height),
+                            &EpochData::new(epoch_randomness.clone()),
+                        )
+                        .log_err()?;
+                }
+            }
+        }
+        BlockStateEventWithIndex::Disconnect(tip_height) => {
+            let disconnected_tip = tip_height.next_height();
+            if chain_config.is_last_block_in_epoch(&disconnected_tip) {
+                db_tx
+                    .del_epoch_data(chain_config.epoch_index_from_height(&disconnected_tip))
+                    .log_err()?;
+            }
+        }
```

Comment:
This makes my head spin somewhat. The if block is entered if the next block after `top_height` is the last block? So the epoch switches at block heights (epoch_length - 1 + K * epoch_length) where K is a non-negative integer, and tip_height refers to the tip after disconnection? That's how I see it making some sense, is it correct? At the very least a comment explaining why this is not an off-by-one (or two?) error would be welcome.

Reply:
thanks for noticing, there is a bug indeed! fixed and covered with tests